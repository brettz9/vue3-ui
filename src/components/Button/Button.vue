<template>
  <button v-bind="$attrs" :class="`${state.styles} py-2 px-6 uppercase transition ease-in-out duration-300`">
    <slot />
  </button>
</template>

<script lang="ts">
import { defineComponent, computed, reactive } from "vue";
import { propsToStyles } from '../../utils';

export default defineComponent({
  name: "Button",
  props: {
    color: {
      type: String,
      required: true,
      default: "wine",
      validator(value:string) {
        return ["wine", "dark", "light"].includes(value);
      }
    },
  },
  setup(props: any, context: any) {
    const state = reactive({
      styles: computed(() => propsToStyles({
        props,
        transform: {
          color: {
            wine: "bg-wine hover:bg-wine-lighter text-white",
            dark: "bg-grey hover:bg-grey-lighter text-white",
            light: "bg-white hover:bg-gray-300 text-black",
          },
        },
      })),
    });
    return { state };
  }
});
</script>
