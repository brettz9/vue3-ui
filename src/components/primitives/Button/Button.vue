<script>
import { computed } from 'vue'

const Button ={
  name: 'VButton',
  props: {
      type: {
          type: String,
          default: 'button'
      },
      size: {
          type: String,
          default: null
      },
      label: {
          type: String,
          default: null,
      },
      rounded: {
          type: Boolean,
          default: false,
      },
      loading: {
        type: Boolean,
        default: false,
      },
      outlined: {
        type: Boolean,
        default: false,
      },
      expanded: {
        type: Boolean,
        default: false,
      },
      inverted: {
        type: Boolean,
        default: false,
      },
      focused: {
        type: Boolean,
        default: false,
      },
      active: {
        type: Boolean,
        default: false,
      },
      hovered: {
        type: Boolean,
        default: false,
      },
      selected: {
        type: Boolean,
        default: false,
      },
      nativeType: {
        type: String,
        default: 'button'
      },
      tag: {
        type: String,
        default: 'button'
      }
  },
  setup(props, { attrs }) {
    const computedTag = computed(() => attrs.disabled ? 'button' : props.tag)
    return { computedTag }
  }
}

export default Button;
</script>

<template>
  <component
    :is="computedTag"
    class="button"
    :type="nativeType"
    v-bind="$attrs"
    :class="[size, type, {
      'is-rounded': rounded,
      'is-loading': loading,
      'is-outlined': outlined,
      'is-fullwidth': expanded,
      'is-inverted': inverted,
      'is-focused': focused,
      'is-active': active,
      'is-hovered': hovered,
      'is-selected': selected
    }]">
    <span v-if="label">{{ label }}</span>
    <slot />
  </component>
</template>
