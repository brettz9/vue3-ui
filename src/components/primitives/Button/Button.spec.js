import { render } from '@testing-library/vue'
import Button from './Button.vue'

let wrapper

describe('Button.vue', () => {
  beforeEach(() => {
    wrapper = render(Button)
  })

  it('is called', () => {
    wrapper.getByRole('button')
  })

  it('render correctly', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })
})
