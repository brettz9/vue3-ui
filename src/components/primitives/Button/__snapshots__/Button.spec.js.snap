// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button.vue render correctly 1`] = `
<button class="button is-primary" type="button">
  <!--v-if-->
</button>
`;
